add a listing of the user's presentations

jberger opened the pull request:
Note that this was not possible when using the routing defined in
a94b60c0f43bf0a840066957d8f7d4a1a16684b4 thus I have had to revert it
in the best way I understand.

jberger committed:
add a listing of the user's presentations

Note that this was not possible when using the routing defined in
a94b60c0f43bf0a840066957d8f7d4a1a16684b4 thus I have had to revert it
in the best way I understand.

jberger commented:
I think the solution to a94b60c0f43bf0a840066957d8f7d4a1a16684b4 is not to nest users under a conference (which we specifically discussed not doing) but rather to define a "current_conference" attribute on the app or some such. It could be detected (we already have that logic) or defined in the configuration (probably best). 

Anyway I'm assigning it to you @jhthorsen , please take a look.

jhthorsen commented:
I think we need to discuss this further on IRC, but I don't think the initial idea about having "/user/profile" will work. The reason why I think "/user/...." is a bad idea is that it's hard to connect the profile page to the conference you're looking at. I see that you suggest using "current_conference" (fetching the "current" conference from the database). The issue is that I don't think this will work for the next conference:
- A user visits http://mct.com/puppy-conference/ and register. All works nicely.
- A couple of months later, the same user register for http://mct.com/mojoconf2015
- The user is redirected to /user/profile, but where will the links in the navbar take the user? To puppy conference or mojoconf2015? And also, which theme will be used? puppy- or mojo-theme?

On the other hand, if we have /:cid/profile the user will still move between the same conference pages. Also, I'm suggesting the profile page has links to the other conferences that is known by the system, so you can move between the pages.

I know, right now this doesn't make much sense, but it's the only thing that makes sense (imo) regarding the data model we decided on, where we have global users and global conferences.

I'm suggesting these routes:
- GET /
  Redirects to the lastest conference. We can make this more clever later on by adding "domain" to the mix, so puppy-conference.com and mojoconf.com can have difference redirects.
- GET /:cid/ (cid = conference ID = mojoconf, puppy-conference, ...)
  Will display the main conference landing page
- GET /:cid/presentations
  Show the presentations for the given conference.
- GET /:cid/presentations/:url_name
  Show one specific presentation. This page can also have an edit button, if the logged in user owns the presentation.
- GET /:cid/profile
  Display the profile page for a user, where you can jump to submitted talks and other conferences in the system.
- GET /:cid/profile/presentations
  Show the presentations for a user. We can sort the conferences so we first have one table with the presentations for the current conference, and then (optionally) for all the other conferences. This page could also show a form for submitting a new presentation.
- POST /:cid/profile/presentations
  Create a new presentation.
- GET/POST /:cid/profile/presentations/:url_name/edit
  Edit a presentation.

(I think I've covered all the routes now)

jberger commented:
First of all, the server is already doing a redirect of `/` to `/<<current_conference>>` so I'm not sure what's so hard about having there be an app default of `current_conference`. If there is a difficulty, it still isn't that hard since they will reach the login page via a specific conference's site, so the act of coming from that site can put a value in the session about which conference they are referring to.

Second, once `/:cid` is removed from the route, the resulting route `GET/POST /user/profile/presentations/:url_name/edit
Edit a presentation.` will not be sufficient. Presentation url_names are not unique wrt the user and indeed many people give the same talk at multiple conferences.

jberger commented:
Personally I don't think there should be a "submit a presentation" action under the user's profile page. As @jhthorsen suggests this is ambiguous, but why is it ambiguous? Because they are unrelated. You should submit a presentation from the conference page.

Links and styling can come either noting the originating site or the server's default conference.

jberger commented:
After discussion on IRC, and seeing that ACT does `/:cid/user/profile` etc I will relent. I'm still not 100% convinced, but it does mean that lots of people disagree with me.

jhthorsen commented:
I agree that a talk should be submitted from /:cid/presentations, and has nothing to do with the profile. The main point here is that I want /:cid/ in front of the routes, so don't change conferences or themeing on "random".

I do think `/:cid/profile/presentations` is a good idea. I'm also fine by replacing "profile" with "user".

Commit:
Change path for profile to /:cid/user/profile

Commit:
Change heading for presentation/edit.html.ep

Commit:
Add $t->$_test_table([....])

Commit:
Add data and tests for user/presentations.html.ep

Commit:
Can't call _will_remove_this_once_prod_is_up_to_date() after a presentation is added

  Can't load application from file "mct/script/mct":
  DBD::Pg::st execute failed: ERROR:  update or delete on table
  "conferences" violates foreign key constraint
  "presentations_conference_fkey" on table "presentations"

jhthorsen commented:
What do you think @jberger ?

jberger commented on lib/MCT/Model/User.pm:
```diff
@@ -45,6 +45,40 @@ sub validate {
+    sub {
+      my ($delay, $err, $results) = @_;
+      die $err if $err;
+      $self->$cb(undef, [map { MCT::Model::Presentation->new(%$_, db => $self->db) } $results->hashes->each]);
```
I originally intentionally didn't load a full object since now we are opening ourselves up for skew between this mechanism of loading presentation objects and the main class.

jhthorsen replied:
That's why we have unit tests.

I partially agree though. The issue is that it's rather confusing to me, when I have to write the templates in two ways: Either the oo-way or the data-structure-way.
